src/psyclone/psyir/nodes/omp_directives.py

```diff
@@ -227,6 +237,814 @@ class OMPSerialDirective(OMPRegionDirective, metaclass=abc.ABCMeta):
+                    # These binary operations are format of Literal MUL Literal
+                    # where step_val is the 2nd literal and the multiplier
+                    # is the first literal
+                    if (not (isinstance(binop.children[0], Literal) and
+                             isinstance(binop.children[1], Literal))):
+                        raise GenerationError(
+                                "Found a dependency index that is a "
+                                "BinaryOperation with a child "
+                                "BinaryOperation with a non-Literal child "
+                                "which is not supported.")
+                    binop_val = int(binop.children[1].value)
+                    num_entries = int(binop.children[0].value)+1
+                elif ref.operator == BinaryOperation.Operator.SUB:
+                    symbol = ref.children[0].symbol
+                    binop = ref.children[1]
+                    if binop.operator != BinaryOperation.Operator.MUL:
+                        raise GenerationError(
+                                "Found a dependency index that is a "
+                                "BinaryOperation with a child "
+                                "BinaryOperation with a non-MUL operand "
+                                "which is not supported.")
+                    # These binary operations are format of Literal MUL Literal
+                    # where step_val is the 2nd literal and the multiplier
+                    # is the first literal
+                    if (not (isinstance(binop.children[0], Literal) and
+                             isinstance(binop.children[1], Literal))):
+                        raise GenerationError(
+                                "Found a dependency index that is a "
+                                "BinaryOperation with a child "
+                                "BinaryOperation with a non-Literal child "
+                                "which is not supported.")
+                    binop_val = -int(binop.children[1].value)
+                    num_entries = int(binop.children[0].value)
+                else:
+                    raise GenerationError("Found a dependency index that is "
+                                          "a BinaryOperation where the "
+                                          "format is Reference OP "
+                                          "BinaryOperation with a non-ADD, "
+                                          "non-SUB operand "
+                                          "which is not supported.")
+            else:
+                raise GenerationError("Found a dependency index that is a "
+                                      "BinaryOperation where neither child "
+                                      "is a Literal or BinaryOperation. "
+                                      "PSyclone can't validate "
+                                      "this dependency.")
+        start = None
```

Comment:
OK, my comments so far was what I noticed while trying to 'reach' this part of the code. To me it feels like this could be replaced with some improvements of the variable access functionality (and I have to admit, I tried to understand the OpenMP standard to see what exactly is happening, but gave up). The variable access would give you all accesses, and you could just 'filter' them.
Some things would be missed, e..g you only allow certain types of BinOp, but with variable access info this would not be caught. But then, aren't these BinOps created by PSyclone in the first place, so is there a reason to check them again here?
If you need the constants of the BinOp, the variable access info should contain a reference to the node in which they are used, so you could get them this way>